File: javascript/section03/nestedLoops.js
// Log all pairs of array
// 1,2,3,4,5
//Eg:
// [1,1] [1,2] [1,3] [1,4] [1,5] 
// [2,1] [2,2]  [2,3] [2,4] [2,5]

const boxes = [1,2,3,4,5];

function logAllPairOfArray(array) {
    for (let i = 0; i < array.length; i++) { //o(n)
        for (let j = 0; j < array.length; j++) { // o(n)
            console.log(array[i],',',array[j]);
        }
    }
}


/*
    The big O of this function would be O(n ˆn) or o(nˆ2)  or quadratic time
    If the loops are nested we multiply them

*/

File: cpp/section03/nestedLoops.cpp
#include <iostream>
#include <vector>

// Function to log all pairs of elements in an array
void logAllPairsOfArray(const std::vector<int>& array) {
    // O(n) - Outer loop
    for (size_t i = 0; i < array.size(); i++) {
        // O(n) - Inner loop
        for (size_t j = 0; j < array.size(); j++) {
            std::cout << array[i] << ", " << array[j] << std::endl;
        }
    }
}

int main() {
    std::vector<int> boxes = {1, 2, 3, 4, 5};

    // Call the function with the boxes vector
    logAllPairsOfArray(boxes);

    return 0;
}

/*
    The overall time complexity is O(n^2) due to the nested loops,
    since we multiply the number of iterations of the two loops.
*/
